File: main-server/src/events/example-group-stream.ts
import { Subscribe } from './index';
import { registerGroup, unregisterGroup, publishGroupEvent } from './group-stream';
import { publish } from '../dal/redis';

// 1. 注册 handler（使用 @Subscribe 装饰器）
class OrderEventHandlers {
    @Subscribe('order.created')
    async handleOrderCreated(data: any) {
        console.log('[GroupStream Example] 收到 order.created 事件:', data);
    }
}


// 2. 注册分组（启动分组消费）
async function startOrderProcessing(groupId: string) {
    await registerGroup('order.created', groupId);
    console.log(`[GroupStream Example] 已注册分组: order.created:${groupId}`);
}

// 3. 取消分组（停止分组消费）
async function stopOrderProcessing(groupId: string) {
    await unregisterGroup('order.created', groupId);
    console.log(`[GroupStream Example] 已注销分组: order.created:${groupId}`);
}

// 4. 发布分组事件
async function createOrder(groupId: string, order: any) {
    await publishGroupEvent('order.created', groupId, order);
    console.log(`[GroupStream Example] 已发布订单创建事件到 group:${groupId}`);
}

// 示例调用
(async () => {
    const groupId = 'group_123';
    await startOrderProcessing(groupId);
    await createOrder(groupId, { orderId: 1, userId: 2 });
    // await stopOrderProcessing(groupId); // 如需注销可取消注释
})();

File: ai-service/app/core/example_group_stream.py
"""
分组顺序消费模式（Group Stream）示例

统一使用 subscribe 注册所有事件处理函数，通过 register_group/unregister_group 管理分组消费状态
"""

from app.core.event_system import (
    get_event_system, 
    register_group,
    unregister_group,
    publish_group_event,
)

# 注册处理函数方式 - 使用 subscribe
# ====================================
async def handle_order_created(data):
    """处理订单创建事件"""
    print(f"收到订单创建事件: {data}")
    # 处理订单...
    return {"status": "processed"}

async def handle_order_refunded(data):
    """处理订单退款事件"""
    print(f"收到订单退款事件: {data}")
    # 处理退款...
    return {"status": "refunded"}

# 注册处理函数
get_event_system().subscribe("order.created", handle_order_created)
get_event_system().subscribe("order.refunded", handle_order_refunded)

# 使用示例
# ====================================
async def start_order_processing():
    """启动订单处理"""
    # 1. 注册分组（对应用户ID为123的订单事件）
    await register_group("order.created", "user_123")
    await register_group("order.refunded", "user_123")
    print("已启动用户123的订单处理")
    
    # 2. 发布订单创建事件
    order_data = {
        "id": "ORD-001",
        "user_id": "user_123",
        "amount": 99.99,
        "items": ["item1", "item2"]
    }
    await publish_group_event("order.created", "user_123", order_data)
    print(f"已发布订单创建事件: {order_data}")
    
    # 稍后可以停止处理
    # await stop_order_processing()
    
async def stop_order_processing():
    """停止订单处理"""
    await unregister_group("order.created", "user_123")
    await unregister_group("order.refunded", "user_123")
    print("已停止用户123的订单处理") 